core/common/src/GenericInstanceFilter.ts

```diff
@@ -0,0 +1,187 @@
+  }
+  export function isPoint3d(value: GenericInstanceFilterRuleValue.Values): value is GenericInstanceFilterRuleValue.Point3d {
+    return isPoint2d(value) && (value as GenericInstanceFilterRuleValue.Point3d).z !== undefined;
+  }
+  export function isInstanceKey(value: GenericInstanceFilterRuleValue.Values): value is GenericInstanceFilterRuleValue.InstanceKey {
+    return (value as GenericInstanceFilterRuleValue.InstanceKey) !== undefined && (value as GenericInstanceFilterRuleValue.InstanceKey).className !== undefined;
+  }
+  export type Values =
+    | string
+    | number
+    | boolean
+    | Date
+    | GenericInstanceFilterRuleValue.Point2d
+    | GenericInstanceFilterRuleValue.Point3d
+    | GenericInstanceFilterRuleValue.InstanceKey;
+}
+
+/**
+ * Defines single filter rule.
+ * @beta
+ */
+export interface GenericInstanceFilterRule {
+  /**
+   * Alias of the source to access this property. For related properties `sourceAlias` should match
+   * `GenericInstanceFilterRelatedInstanceDescription.alias` of one `GenericInstanceFilter.relatedInstances`.
+   */
+  sourceAlias: string;
+  /**
+   * Property name for accessing property value.
+   */
+  propertyName: string;
+  /**
+   * Comparison operator that should be used to compare property value.
+   */
+  operator: GenericInstanceFilterRuleOperator;
+  /**
+   * Value to which property values is compared to. For unary operators value is 'undefined'.
+   */
+  value?: GenericInstanceFilterRuleValue;
+  /**
+   * Type name of the property.
+   */
+  propertyTypeName: string;
+}
+
+/**
+ * Type definition that describes operators supported by `GenericInstanceFilterRuleGroup`.
+ * @beta
+ */
+export type GenericInstanceFilterRuleGroupOperator = "and" | "or";
+
+/**
+ * Group of filter rules joined by logical operator.
+ * @beta
+ */
+export interface GenericInstanceFilterRuleGroup {
+  /**
+   * Operator that should be used to join rules.
+   */
+  operator: GenericInstanceFilterRuleGroupOperator;
+  /**
+   * List of rules or rule groups that should be joined by `operator`.
+   */
+  rules: Array<GenericInstanceFilterRule | GenericInstanceFilterRuleGroup>;
+}
+
+/**
+ * Describes related instance whose property was used in the filter.
+ * @beta
+ */
+export interface GenericInstanceFilterRelatedInstanceDescription {
+  /**
+   * Describes path that should be used to reach related instance from the source.
+   */
+  path: GenericInstanceFilterRelationshipStep[];
+  /**
+   * Related instance alias. This alias match `sourceAlias` in all filter rules where
+   * properties of this related instance is used.
+   */
+  alias: string;
+}
+
+/**
+ * Describes single step between source class and target class.
+ * @beta
+ */
+export interface GenericInstanceFilterRelationshipStep {
+  /** Class name of the source class. */
+  sourceClassName: string;
```

Comment:
the comment for class name properties should be more specific about the format

```suggestion
  /** Full class name of the source class, e.g. `BisCore.Element`. */
  sourceClassName: string;
```

Reply:
Updated comments on other props.

Reply:
Just to confirm, do we accept only `.`, or `:` is also fine?

Reply:
Actually we accept only `:` as we use class names that we get from content descriptor.